### PR TITLE
Add goerli support

### PR DESCRIPTION
**Pull request by holgerd77**
Closes https://github.com/ethereumjs/ethereumjs-common/issues/29.

### CODE REVIEW

**Comment by holgerd77** [comment]
I think once this is generally ok, we should rather quickly merge and do a minor or so release, just to be able to start experimenting with this a bit.

Then we can on top have discussions around changes on the genesis file format. My first (second thought :-)) impression here is also that we should iterate quickly, since we do lot of things different here and we should probably quickly start changes the ones obviously adoptable since this will also ease further discussion around this and avoid confusion and getting lost by not seeing the pattern left for a substantial discussion.

File: tests/chains.js [hunk 1]
```diff
@@ -57,6 +57,8 @@ tape('[Common]: Initialization / Chain params', function (t) {
     st.equal(c.genesis().hash, '0x6341fd3daf94b748c72ced5a5b26028f2474f5f00d824504e4fa37a75767e177', 'rinkeby')
     c.setChain('kovan')
     st.equal(c.genesis().hash, '0xa3c565fc15c7478862d50ccd6561e3c06b24cc509bf388941c25ea985ce32cb9', 'kovan')
+    c.setChain('goerli')
+    st.equal(c.genesis().timestamp, '0x5bdda800', 'goerli')
```
Comment (axic) [inline tests/chains.js]: Why not check the hash?

File: chains/goerli.json [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+    "difficulty": 1,
+    "nonce": "0x0000000000000000",
+    "extraData": "0x2249276d20646f6e652077616974696e672e2e2e20666f7220626c6f636b2066696e616c69747922202d2049676779270000000001fa1804c408085d9c57eeb167ce953c99b6cb1e20794Fd02933F303FbA550bd1fe2f0649E3576eB0000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000",
+    "stateRoot": ""
+  },
+  "hardforks": [
+    {
+      "name": "chainstart",
+      "block": 0,
+      "consensus": "poa",
```
Comment (axic) [inline chains/goerli.json]: I think this should say `clique` instead.

Comment (vpulim) [reply chains/goerli.json]: agree. we should also modify ``chains/rinkeby.json`` to use ``"consensus": "clique"``

Comment (holgerd77) [reply chains/goerli.json]: Yes, also agree. I would nevertheless like to for now keep within this PR and just put the goerli config into the existing format for early experimentation and make these kind of updates part of a second round of larger changes and generally rework our terminology here to at least come closer to a common genesis file format, see https://github.com/ethereumjs/ethereumjs-common/issues/30.

File: tests/chains.js (hunk 1 above)
Comment (holgerd77) [reply tests/chains.js]: There was only ``mixHash`` provided in the Goerli [genesis specification(s)](https://github.com/goerli/testnet/blob/master/geth/goerli.genesis) and I wasn't sure how this relates to the generic ``hash`` field here, so I wanted to take a parameter which is disambiguous.

**Comment by holgerd77** [comment]
My idea here for going on would be:

1. Open a PR from the client library including this PR as a dependency and see if we can connect to Goerli, I can do this in the coming days, but @vpulim feel also free if you are interested or anyone else
2. Do a minor release here
3. Further experiment with Goerli, let thinks "ripen"
3. Articulate some of our needs and ideas for a common genesis format over on the [EIP](https://github.com/ethereum/EIPs/issues/1085)
4. Integrate here as much as possible, do a breaking release

**Comment by vpulim** [comment]
@holgerd77 Before we can connect to the goerli testnet, the ``hash`` and ``stateRoot`` fields must be correctly filled in. The client can compute these via the ``--params`` flag from a geth genesis state file. I'll try to grab these values later today and post back.

File: tests/chains.js (hunk 1 above)
Comment (vpulim) [reply tests/chains.js]: The genesis spec formats that Geth and Parity use don't explicitly specify the hash (and state root). It is usually computed by the client, but in our case, we meed to compute it offline and enter it into the spec.

File: chains/goerli.json [hunk 3]
```diff
@@ -0,0 +1,145 @@
+{
+  "name": "goerli",
+  "chainId": 6284,
+  "networkId": 6284,
+  "comment": "Cross-client PoA test network",
+  "url": "https://github.com/goerli/testnet",
+  "genesis": {
+    "hash": "0x0000000000000000000000000000000000000000000000000000000000000000",
```
Comment (vpulim) [inline chains/goerli.json]:
```suggestion
    "hash": "0xfa57319d09fd8a32faaf18d338c8a925a5a7975285bf29ecd024e083cba8abb1",
```

File: chains/goerli.json [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+    "timestamp": "0x5bdda800",
+    "gasLimit": 10485760,
+    "difficulty": 1,
+    "nonce": "0x0000000000000000",
+    "extraData": "0x2249276d20646f6e652077616974696e672e2e2e20666f7220626c6f636b2066696e616c69747922202d2049676779270000000001fa1804c408085d9c57eeb167ce953c99b6cb1e20794Fd02933F303FbA550bd1fe2f0649E3576eB0000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000",
+    "stateRoot": ""
```
Comment (vpulim) [inline chains/goerli.json]:
```suggestion
    "stateRoot": "0x0f410b6a6eae3d3156eccd966ac842a4c545c47921b9fe36386de18152cfddcf"
```

**Comment by holgerd77** [comment]
Ok, did all the requested changes respectively accepted the suggestions, squashed the commits and force-pushed.

**Comment by holgerd77** [comment]
@vpulim Will you do a live-test of this (or have you already)?

**Comment by vpulim** [comment]
Blocks are downloading!

```
INFO [11-26|15:01:20] Data directory: goerli/goerli/ethereumjs/chaindata
INFO [11-26|15:01:20] Initializing Ethereumjs client...
INFO [11-26|15:01:20] Fast sync mode
INFO [11-26|15:01:20] Connecting to network: goerli
INFO [11-26|15:01:20] Latest local block: number=0 td=1 hash=fa57319d...
INFO [11-26|15:01:20] Synchronizing blockchain...
INFO [11-26|15:01:20] Listener up transport=rlpx url=enode://010e1eacedb6f82c8d3819cdf34fbc46d73391925bd5f281aa7310b340986ed8f0692aadf5d467f0f264872ee2b1314cfb1ce1b1bd5bbcc39dac8454520239e8@[::]:30303
INFO [11-26|15:01:23] Listener up transport=libp2p url=/ip4/127.0.0.1/tcp/50505/ipfs/QmeRX39Y4sxbcwN4F1BGntUP9ff3kvJmEiq1ReNUT1Ecqv
INFO [11-26|15:01:23] Listener up transport=libp2p url=/ip4/127.0.0.1/tcp/50580/ws/ipfs/QmeRX39Y4sxbcwN4F1BGntUP9ff3kvJmEiq1ReNUT1Ecqv
INFO [11-26|15:01:23] Found fast peer: id=573b6607 address=52.56.136.200:30303 transport=rlpx protocols=eth
INFO [11-26|15:01:23] Found fast peer: id=46add44b address=94.237.54.114:30313 transport=rlpx protocols=eth
INFO [11-26|15:01:32] Using origin peer: id=573b6607cd59f241e30e4c4943fd50e99e2b6f42f9bd5ca111659d309c06741247f4f1e93843ad3e8c8c18b6e2d94c161b7ef67479b3938780a97134b618b5ce address=52.56.136.200:30303 transport=rlpx protocols=eth height=92921
INFO [11-26|15:01:33] Imported blocks count=128 number=1 hash=569f8f70...
INFO [11-26|15:01:34] Imported blocks count=128 number=129 hash=56141179...
INFO [11-26|15:01:35] Imported blocks count=128 number=257 hash=9776dec1...
INFO [11-26|15:01:35] Imported blocks count=128 number=385 hash=a9c41745...
INFO [11-26|15:01:36] Imported blocks count=128 number=513 hash=5ff28f69...
INFO [11-26|15:01:37] Imported blocks count=128 number=641 hash=694aab6c...
INFO [11-26|15:01:37] Imported blocks count=128 number=769 hash=512ec5ec...
INFO [11-26|15:01:38] Imported blocks count=128 number=897 hash=2476677a...
INFO [11-26|15:01:38] Imported blocks count=128 number=1025 hash=15ee9150...
```

**Comment by vpulim** [comment]
Also, submitted a new issue during the process of calculating the stateRoot: https://github.com/ethereumjs/ethereumjs-common/issues/32

**Comment by 5chdn** [comment]
Cool!

What genesis format are you using here? 

Also, please note that we are going to reset the testnet as soon as Parity integration is done (we are still in pre-testnet stage).

**Comment by vpulim** [comment]
@5chdn ``ethereumjs/ethereumjs-common`` uses a custom JSON format for encoding genesis params. However, ``ethereumjs/ethereumjs-client`` can also load a geth genesis format file via the ``--params`` option. But since this format doesn't include bootnodes, they must be specified via an additional parameter.

I am planning to also support parity's chain spec format in the future since I believe it is the easiest to use from an end-user perspective (requires specifying only a single file to load genesis params, bootnodes and initial state), unless EIP-1085 ends up incorporating all of these as well.

**Comment by holgerd77** [comment]
@5chdn Ah sorry, very likely not, no one really closely followed apart from happily taking note 😄, will re-check on the configuration during the next days.

**Pull request merged**